agent/crates/public/src/segment_map.rs

```diff
@@ -0,0 +1,251 @@
+            } else {
+                events.push((u16::MAX, false, idx));
+            }
+        }
+
+        events.sort_by(|a, b| a.0.cmp(&b.0).then_with(|| b.1.cmp(&a.1)));
+
+        let curr_len = self.segments.len();
+        let mut resort_segments = Vec::with_capacity(curr_len);
+        let mut active_segments = vec![false; curr_len];
+        let mut last_pos = 0u16;
+
+        for (pos, is_start, seg_idx) in events {
+            if pos > last_pos {
+                // 所有活跃的 events 对应的 value 合集，即为新段的 value
+                // all active events value concat as new value for new segments
+                let all_start_segments_value: Vec<T> = active_segments
+                    .iter()
+                    .enumerate()
+                    .filter(|&(_, &active)| active)
+                    .flat_map(|(idx, _)| self.segments[idx].value.iter())
+                    .copied()
+                    .collect();
+                if !all_start_segments_value.is_empty() {
+                    resort_segments.push(Segment::new(
+                        last_pos,
+                        (pos - 1).max(u16::MAX),
+                        all_start_segments_value,
+                    ));
+                }
+            }
+
+            // when meet end port, the segment is no longer active
+            active_segments[seg_idx] = is_start;
+            last_pos = pos;
+        }
+        self.segments = resort_segments;
+        self.dedup_segments();
+    }
+
+    pub fn find(&self, port: u16) -> Option<Vec<T>> {
+        let mut left = 0;
+        let mut right = self.segments.len();
+
+        while left < right {
+            let mid = left + (right - left) / 2;
+            let segment = &self.segments[mid];
+
+            if port < segment.start {
+                right = mid;
+            } else if port > segment.end {
+                left = mid + 1;
+            } else {
+                return Some(self.segments[mid].value.clone());
+            }
+        }
+        None
+    }
+}
+
+impl<T: Clone + Debug> Display for SegmentMap<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        for segment in &self.segments {
+            writeln!(f, "{}-{}: {:?}", segment.start, segment.end, segment.value)?;
+        }
+        Ok(())
+    }
+}
+
+pub fn parse_u16_range_list_to_port_pairs(
+    port_str: impl AsRef<str>,
+    strict: bool,
+) -> Option<Vec<(u16, u16)>> {
+    let port_str = port_str.as_ref();
+    let mut port_pair = Vec::new();
+    let mut ports = port_str.split(",");
+    while let Some(mut p) = ports.next() {
+        p = p.trim();
+        if let Ok(port) = p.parse::<u16>() {
+            port_pair.push((port, port));
+        } else {
+            let range = p.split("-").collect::<Vec<&str>>();
+            if range.len() != 2 {
+                if strict {
+                    return None;
+                }
+                continue;
+            }
```

Comment:
```suggestion
            let mut range = p.split("-");
            let (start, end) = (range.next(), range.next());
            if end.is_none() || !range.next().is_none()
                if strict {
                    return None;
                }
                continue;
            }
```
下面可以直接 start.unwrap end.unwrap